File: cppStuff/interview/Sorting_Bubble_Sort.cpp
/*
Given an n element array A = a0, a1,...,a(n-1) fo distint elements, sort array A in ascending order using the 
Bubble Sort algorithm above. Once sorted, print the following three line: 
	1. Arrays is sorted in numSwaps swaps, where numSwaps is the number of swaps that took place
	2.  First Element: firstElement, where  firstElement is the first element in the sorted array
	3. Last Element: lastElement, where  lastElement is the last element in the sorted array
*/

#include <vector>
#include <iostream>

void swap(std::vector<int> & a, int j)
{
    int temp = a[j-1];
    a[j-1] = a[j];
    a[j] = temp;
}

int bubbleSort(std::vector<int> & a)
{
    int count = 0;
    for(int i = 1; i < a.size(); i++)
    {
        for(int j = 1; j < a.size(); j++)
        {
            if(a[j-1] > a[j])
            {
                swap(a, j);
                count++;
            }
        }
    }
    return count;
}

int main()
{
    int n;
    std::cin >> n;
    std::vector<int> a(n);
    for(int a_i = 0;a_i < n;a_i++)
       std::cin >> a[a_i];
    int numberOfSwaps = bubbleSort(a);
    std::cout << "Array is sorted in " <<  numberOfSwaps << " swaps.\n";
    std::cout << "First Element: " << a[0] << '\n';
    std::cout << "Last Element: " << a[n-1] << '\n';
    return 0;
}

File: pStuff/SortingBubbleSort.py
#!/bin/python
'''
Given an n element array A = a0, a1,...,a(n-1) fo distint elements, sort array A in ascending order using the 
Bubble Sort algorithm above. Once sorted, print the following three line: 
	1. Arrays is sorted in numSwaps swaps, where numSwaps is the number of swaps that took place
	2.  First Element: firstElement, where  firstElement is the first element in the sorted array
	3. Last Element: lastElement, where  lastElement is the last element in the sorted array
'''

import sys

def bubbleSort(a) : 
    numberOfSwaps = 0
    for i in range(1, len(a)) : 
        for j in range(1, len(a)) : 
            if( a[j-1] > a[j] ) : 
                a[j-1], a[j] = a[j], a[j -1]
                numberOfSwaps += 1
    return numberOfSwaps

def main() : 
    n = int(raw_input().strip())
    a = map(int, raw_input().strip().split(' '))
    numberOfSwaps = bubbleSort(a)
    print "Array is sorted in %d swaps." % numberOfSwaps
    print "First Element: %d" % a[0]
    print "Last Element: %d" % a[n-1]

if __name__ == "__main__" : 
    main()
